### ISSUE TITLE
SHFB builds for UWP Windows 10 Fall Creators Update (10.0; Build 16299) produce unexpected content

### ISSUE THREAD

**Issue by dkemlage** [EWSoftware/SHFB#581]
Hello Eric,

We are moving our Universal Windows Platform (UWP) API to target the new **Windows 10 Fall Creators Update (10.0; Build 16299)** and are finding that SandCastle VSIX extension is producing undesirable help documentation results. We were previously targeting the UWP **Windows 10 Anniversary Edition (10.0, Build 14393)** version and the SandCastle generated output was great. 

Here is the configuration of our build machine:

Windows 10 version 1709 (OS Build 16299.125) | Visual Studio Enterprise 2017 version 15.5.7 | SHFB 2017.12.30.2
 
Here are the steps to reproduce the issue:

**ONE:** Use VS2017 to create a Windows Universal Class Library. In the Project Properties dialog, select the Library tab and ensure that both the ‘Target version’ and ‘Min version’ combo box drop downs are set to **Windows 10 Anniversary Edition (10.0, Build 14393)**. Also don’t forget to select the Build tab and check on the ‘XML documentation file’ checkbox. 

![image](https://user-images.githubusercontent.com/5495810/37128638-e6adc43e-222f-11e8-9bb2-df4c9b270fdb.png)

**TWO:** Replace the code in the Class1.cs file with the following:
```
using System.Threading.Tasks;

namespace TEST_DLL
{

    /// <summary>
    /// Represents a test type to check sand castle output.
    /// </summary>
    public class Test
    {
        /// <summary>
        /// Asynchronously adds a new attachment.
        /// </summary>
        /// <param name="name">The name for the attachment.</param>
        /// <param name="contentType">The content type for the attachment.</param>
        /// <param name="data">The data for the attachment.</param>
        /// <returns>
        /// A task that represents the asynchronous add attachment operation. The value of the task result is an <see cref="Attachment"/> object.
        /// </returns>
        public Task<Attachment> AddAttachmentAsync(string name, string contentType, byte[] data)
        {
            return Task.FromResult(new Attachment());
        }

    }

    /// <summary>
    /// Represents an attachment type.
    /// </summary>
    public sealed class Attachment
    {
    }

}
```
**THREE:** Build the VS2017 project to create the TEST_DLL.dll and TEST_DLL.xml files.

**FOUR:** Create a VS2017 SandCastle documentation project. Reference the TEST_DLL.dll and TEST_DLL.xml files. Make sure the Framework version is set to **.NET for Universal Windows** in the documentation project's Properties dialog | Build tab.

**FIVE:** Build the SandCastle project and examine the resulting .chm file for the AddAttachmentAsync Method. As you can see the documentation looks great and the signature for the Test.AddAttachmentAsync Method is correct: 
```
public Task<Attachment> AddAttachment(string name, string contentType, byte[] data).
```
![image](https://user-images.githubusercontent.com/5495810/37128689-24f9c058-2230-11e8-9bfc-2baa9eddb2af.png)

Now to see the problem when targeting the UWP **Windows 10 Fall Creators Update (10.0; Build 16299)** version of the builds, do the following:

**SIX:** Go back to the VS2017 Windows Universal Class Library project. In the Project Properties dialog, select the Library tab and change both the ‘Target version’ and ‘Min version’ combo box drop downs to **Windows 10 Fall Creators Update (10.0; Build 16299)**.

![image](https://user-images.githubusercontent.com/5495810/37128727-44db1da4-2230-11e8-8a80-19224ae9bb42.png)

**SEVEN:** Build the VS2017 project to create the TEST_DLL.dll and TEST_DLL.xml files.

**EIGHT:** Go back to the VS2017 SandCastle documentation project and reference the new versions of the TEST_DLL.dll and TEST_DLL.xml files that were created with the **Windows 10 Fall Creators Update (10.0; Build 16299)** version build. 

**NINE:** Build the SandCastle project and examine the resulting .chm file for the AddAttachmentAsync Method. As you will see the documentation is wrong with the signature of the Test.AddAttachment Method not matching was defined in the .cs file or the resulting TEST_DLL.xml. The resulting .chm states the Type signature is: 
```
public Task AddAttachment(void name, Attachment contentType, string data) 
```
which is incorrect.

![image](https://user-images.githubusercontent.com/5495810/37128783-7d5be6fe-2230-11e8-8043-3f1b9015364a.png)

Can you please confirm if this is an issue that needs to be fixed or tell me if I am doing something incorrectly in producing Universal Windows Platform (UWP) Library Reference documentation,

Thanks,
Don Kemlage
dkemlage@esri.com

**Comment by EWSoftware** [comment]
That issue is typically caused by an incorrect framework version setting in the SHFB project.  You might try the ".NET For Universal Windows (NuGet)" option which may resolve it.  You might also try the "Cross Platform" option.  If they don't work, let me know and I'll look into it when I get a chance this week.

**Comment by dkemlage** [comment]
Hi Eric,

I have bitter-sweet news to report.

Compiling the SandCastle documentation project using the Framework version of **.NET For Universal Windows (NuGet)** did not work for my simple use case (TEST_DLL.dll). The same erroneous results in the .chm output remained.

Compiling the SandCastle documentation project using the Framework version of **Cross-platform (.NET Core/.NET Standard)** DID WORK for my simple use case (TEST_DLL.dll), BUT in reality we have a much more complex UWP based .dll to generate Library Reference documentation for. When I tried to build the SandCastle documentation project on our more complex .dll using the Framework version of **Cross-platform (.NET Core/.NET Standard)**, the SandCastle documentation project build threw an error. Here is the relevant portion of the **LastBuild.log** file that should shed some additional light on the issue:

```
  MRefBuilder (v2017.12.30.2)
  Copyright c 2006-2017, Microsoft Corporation, All Rights Reserved.
  Portions Copyright c 2006-2016, Eric Woodruff, All Rights Reserved.
  No source code context base path specified.  Source context information is unavailable.
  Loaded 2 assemblies for reflection and 1 dependency assemblies.
MRefBuilder : warning : Ignoring unresolved assembly reference: System.ObjectModel (System.ObjectModel, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a) required by Esri.ArcGISRuntime [C:\applications\output\win_net_sdk\tools\api-ref\Website\Pass1\UWP\Help\Working\GenerateRefInfo.proj]
MRefBuilder : error : Unresolved assembly reference: Windows.Foundation.UniversalApiContract (Windows.Foundation.UniversalApiContract, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null) required by Esri.ArcGISRuntime [C:\applications\output\win_net_sdk\tools\api-ref\Website\Pass1\UWP\Help\Working\GenerateRefInfo.proj]
    Last step completed in 00:00:00.9843
</buildStep>
<buildStep step="Failed">

SHFB: Error BE0043: Unexpected error detected in last build step.  See build log for details.
   at SandcastleBuilder.Utils.BuildEngine.TaskRunner.Run(String processFilename, String targetFile, String arguments)
   at SandcastleBuilder.Utils.BuildEngine.TaskRunner.RunProject(String projectFile, Boolean minimalOutput)
   at SandcastleBuilder.Utils.BuildEngine.BuildProcess.Build()
```
I look forward to any more suggestions you may have.

Thanks,
-don

**Comment by EWSoftware** [comment]
You can try the assembly binding redirection plug-in.  If you know the version of the reference you are using, you can redirect the unresolved version to it.  If you don't or can't figure it out, you can always add the assembly name to the ignored list.  That's not ideal and it won't include inherited members or documentation for the missing inherited members of classes deriving from there if you have any but it should allow the build to continue.

**Comment by dkemlage** [comment]
Hi Eric,

I could not get the **Binding Redirections** tab of the 'Configure Assembly Binding Redirection Plug-In' to work, SandCastle kept throwing build errors.

I was able to use the **Ignore If Unresolved** tab of the 'Configure Assembly Binding Redirection Plug-In' to get the SandCastle documentation project to compile. However as you pointed out, it is not ideal given that numerous hyperlinks in the documentaion pages are now just **BOLDED** Type names. 
The Microsoft Assemblies that SandCastle could not resolve were:
Windows.Foundation.UniversalApiContract (Windows.Foundation.UniversalApiContract, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null)
Windows.Foundation.FoundationContract (Windows.Foundation.FoundationContract, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null)

![image](https://user-images.githubusercontent.com/5495810/37498466-b9f5554a-287b-11e8-990d-47f428d9c8f7.png)

So I guess my next set of questions/comments are as follows:

For the **Binding Redirections** tab of the 'Configure Assembly Binding Redirection Plug-In', I tried setting the following parameters:
Assembly Name : Windows.Foundation.UniversalApiContract
OldVersion: 5.0.0.0
NewVersion: 4.0.0.0 (I also tried 3.0.0.0 and 2.0.0.0, with no luck)
![image](https://user-images.githubusercontent.com/5495810/37498766-5b3c5272-287d-11e8-8376-42c043beea1e.png)
Did I theoretically do this correctly?

And if I did use the **Binding Redirections** tab correctly is there something else that I need to do or install on my system so that SandCastle can  correctly find the Windows.Foundation.UniversalApiContract  Version=5.0.0.0 and Windows.Foundation.FoundationContract Version=3.0.0.0 Assemblies?

I did find a Microsoft document that talks about which Assemblies are installed as part of the various Windows 10 operating systems (https://docs.microsoft.com/en-us/uwp/extension-sdks/windows-universal-sdk). I do have SandCastle installed on a box with Windows 10 Fall Creators Update (10.0; Build 16299) which according to the Microsoft document should have these versions of the Assemblies installed. Does SandCastle not know how to find these Assemblies on my system?

I look forward to your next reply.

Thanks,
-don

**Comment by EWSoftware** [comment]
Frameworks used to be so much simpler.  Everything was in the Reference Assemblies folder.  That's no longer the case with various frameworks in the Reference Assemblies folder and SDK folders and newer ones like .NETCore and .NETStandard and various extensions to them distributed as NuGet packages as well.  That makes it more difficult to find the right set of assemblies for anything but the base .NET framework.

That assembly name does exist in the reflection data set files for the Universal Windows set but I only see version 1.0, 2.0, and 3.0.   Looking in the related SDK folder I don't see anything above 3.0.  Where 5.0 is at, I have no idea.  If you can find it, let me know and if it's in an SDK folder, I can probably update the reflection data set files to include it in a future release.

The Cross Platform option is basically a redirect to one of the base .NET Framework reflection data sets based on the rules I could find.  Assemblies like that one causing the problem are assumed to come from a NuGet package so I have to rely on what I can find in the documentation source projects.  I suppose it's possible there are multiple versions in the NuGet package and it's not picking up the right one.

Best case for now, if you can find the actual assembly with the version you're looking, is to add it as a reference to the SHFB project.  That should override anything else it finds when looking at the documentation source project references including the NuGet packages from them.  Anything it finds and uses as a reference is listed in the build log.

**Comment by dkemlage** [comment]
Hello Eric,

I have made some progress but I am still not quite all the way there.  I talked to our lead developer and I was able to locate the  **Windows.Foundation.UniversalApiContract Version=5.0.0.0** and **Windows.Foundation.FoundationContract Version=3.0.0.0** Assemblies on my system. They are located at:
```
C:\Program Files (x86)\Windows Kits\10\References\10.0.16299.0\Windows.Foundation.UniversalApiContract\5.0.0.0\Windows.Foundation.UniversalApiContract.winmd
C:\Program Files (x86)\Windows Kits\10\References\10.0.16299.0\Windows.Foundation.FoundationContract\3.0.0.0\Windows.Foundation.FoundationContract.winmd
```
Now I am able to get the SandCastle project to compile without having to use the **Ignore If Unresolved** tab of the **Configure Assembly Binding Redirection Plug-In**.

I did get a weird dialog when trying to add one of the .winmd files as a Reference to the SandCastle project:
![image](https://user-images.githubusercontent.com/5495810/38648182-c0b1f878-3da4-11e8-950e-06566178ab96.png)

I have the **Framework version** set to **Cross-platform (.NET Core/.NET Standard)** as you had suggested. Neither **.NET for Universal Windows** or **.NET for Universal Windows (NuGet)** would compile successfully.

In the SandCastle build log I am getting numerous build warnings for **BuildAssembler : warning : ResolveReferenceLinksComponent:** issues. Unfortunately, I do not know what .dlls to add as a Reference to the SandCastle project - no help coming from the build.log file. These are the **Unknown reference link target** that I am trying to deal with:
```
T:Windows.Security.Credentials.UI.CredentialPicker
T:Windows.Security.Authentication.Web.WebAuthenticationBroker
T:Windows.Security.Credentials.PasswordVault
T:Windows.Security.Credentials.UI.CredentialSaveOption
M:Windows.UI.Xaml.FrameworkElement.ArrangeOverride(Windows.Foundation.Size)
T:Windows.UI.Xaml.Media.ImageSource
T:Windows.Foundation.Point
T:Windows.Foundation.Size
T:Windows.UI.Xaml.DependencyObject
T:Windows.UI.Xaml.UIElement
T:Windows.UI.Xaml.FrameworkElement
T:Windows.UI.Xaml.Controls.Control
T:Windows.UI.Xaml.Controls.ContentControl
P:Windows.UI.Xaml.Controls.ContentControl.Content
P:Windows.UI.Xaml.Controls.ContentControl.ContentTemplate
P:Windows.UI.Xaml.Controls.ContentControl.ContentTemplateRoot
P:Windows.UI.Xaml.Controls.ContentControl.ContentTemplateSelector
P:Windows.UI.Xaml.Controls.ContentControl.ContentTransitions
M:Windows.UI.Xaml.FrameworkElement.OnApplyTemplate
M:Windows.UI.Xaml.Controls.ContentControl.OnContentChanged(System.Object,System.Object)
M:Windows.UI.Xaml.Controls.ContentControl.OnContentTemplateChanged(Windows.UI.Xaml.DataTemplate,Windows.UI.Xaml.DataTemplate)
M:Windows.UI.Xaml.Controls.ContentControl.OnContentTemplateSelectorChanged(Windows.UI.Xaml.Controls.DataTemplateSelector,Windows.UI.Xaml.Controls.DataTemplateSelector)
T:Windows.UI.Xaml.DependencyObject
T:Windows.UI.Xaml.DependencyProperty
M:Windows.UI.Xaml.Controls.Control.OnLostFocus(Windows.UI.Xaml.RoutedEventArgs)
T:Windows.UI.Xaml.Thickness
T:Windows.UI.Xaml.Automation.Peers.AutomationPeer
T:Windows.UI.Xaml.Automation.Peers.FrameworkElementAutomationPeer
```
I am open to any additional suggestions that you may have.

Thanks,
-don

**Comment by EWSoftware** [comment]
SHFB projects don't declare a framework version so I think the warning dialog is normal and you can just ignore it and select Yes.

Regarding the unknown reference link targets, I found references to all of them in the Universal Windows reflection data sets.  Those data sets were generated from the following locations:
- %ProgramFiles(x86)%\Microsoft SDKs\NETCoreSDK
- %ProgramFiles(x86)%\Windows Kits\10\References
- %ProgramFiles(x86)%\Windows Kits\8.1\References\CommonConfiguration\Neutral
- %ProgramFiles(x86)%\Microsoft SDKs\NuGetPackages

**Comment by aceataldot** [comment]
I'm having an issue with similar output that's in issue #620 Nullable parameters not generating good help file entries.

In my case, it's a .NET Standard 2.0 project and I've already set the framework version to "Cross-Platform (.Net Core/.NET Standard).

My build is completing successfully.

I think these might be related since the symptoms are the same. 

Please see the issue #620 for more details.  (If I find a solution to it, I may post it here.)

**Comment by aceataldot** [comment]
In #620, it was an unresolved netstandard.dll.  I had to manually add a reference to a copy of that.  That doesn't see right.

**Comment by EWSoftware** [comment]
I think I have resolved the initial problem with this issue related to the parameter types showing as "void" when using a later release of the targeting version.  It was caused by problems in how SHFB was including mscorlib.dll, netstandard.dll, and System.Runtime.dll as references.  This will be fixed in the next release due out soon.

One issue remains and that is the missing reference to Windows.Foundation.UniversalApiContract and others like it.  They still cannot be found.  They just don't appear as a reference in the projects or in the NuGet package references.  I suspect that they are probably a dependency referenced indirectly by one of the other reference assemblies.  They don't appear to be in the GAC either.  The only solutions are to find a copy and include them as a reference in the SHFB project or add the Assembly Binding Redirection plug-in and add them as one of the ignored references if not found.
